A78-grpc-metrics-wrr-pf-xds.md

```diff
@@ -0,0 +1,154 @@
+
+| Name        | Disposition | Description |
+| ----------- | ----------- | ----------- |
+| grpc.target | required | Indicates the target of the gRPC channel in which PF is used.  (Same as the attribute defined in [A66].) |
+
+The following metrics will be exported:
+
+| Name          | Type  | Unit  | Labels  | Description |
+| ------------- | ----- | ----- | ------- | ----------- |
+| grpc.lb.pick_first.disconnections | Counter | {disconnection} | grpc.target | Number of times the selected subchannel becomes disconnected. |
+| grpc.lb.pick_first.connection_attempts_succeeded | Counter | {attempt} | grpc.target | Number of successful connection attempts. |
+| grpc.lb.pick_first.connection_attempts_failed | Counter | {attempt} | grpc.target | Number of failed connection attempts. |
+
+### XdsClient
+
+The XdsClient component was originally described in [A27].  Note that in
+[A71], we are moving from a single global XdsClient instance to a
+separate global XdsClient instance for each channel target.  The
+proposed metric schema here reflects that change.
+
+XdsClient metrics will have the following labels:
+
+| Name        | Disposition | Description |
+| ----------- | ----------- | ----------- |
+| grpc.target | required | For clients, indicates the target of the gRPC channel in which the XdsClient is used (i.e., the same as the attribute defined in [A66]). For servers, will be the string "#server". |
+| grpc.xds.server | required | The target URI of the xDS server with which the XdsClient is communicating. |
+| grpc.xds.authority | required | The xDS authority.  The value will be "#old" for old-style non-xdstp resource names. |
+| grpc.xds.cache_state | required | Indicates the cache state of an xDS resource.  The value will be one of: <ul><li>"requested": The resource has been requested from the xDS server but has not yet been received.<li>"does_not_exist": The server has indicated that the resource does not exist.<li>"acked": The resource has been received and is valid.<li>"nacked": The resource was received but was not valid.<li>"nacked_but_cached": There is a version of the resource cached, but the most recent update of the resource was invalid.</ul> |
+| grpc.xds.resource_type | required | Indicates an xDS resource type, such as "envoy.config.listener.v3.Listener". |
+
+The following metrics will be exported:
+
+| Name          | Type  | Unit  | Labels  | Description |
+| ------------- | ----- | ----- | ------- | ----------- |
+| grpc.xds_client.connected | Gauge | {bool} | grpc.target, grpc.xds.server | Whether or not the xDS client currently has a working ADS stream to the xDS server.  For a given server, this will be set to 1 when the stream is initially created.  It will be set to 0 when we have a connectivity failure or when the ADS stream fails without seeing a response message, as per [A57].  Once set to 0, it will be reset to 1 when we receive the first response on an ADS stream. |
```

Comment:
I think we will be able to tell that by seeing the backup server listed in this metric, so it doesn't seem necessary to add an additional one.

Reply:
I think a counter type could provide a more reliable signal on the control plane availability, it may be worth it to add separately.

IMO gauge is good to watch trends etc but is unreliable to track occurrences of an event since it loses what happened between snapshots collected. I expect monitoring client TD fail over is one common use case, then a gauge can be ineffective in monitoring brief control plane unavailability events.

Reply:
How likely is it that we'll be using the fallback server for a shorter period of time than the metric collection interval?  If we think that will be a common enough occuurance, then I agree that a counter may be useful.  But if not, the counter won't really add much value.

Reply:
I think an outage is more likely last longer than the collection period but not always, and we may never know the extend of short outages if they happen. In prod the time granularity is configurable, AFAIK 1 minute is the shortest and the default.

One example we have is micro outages we detected from a swarm of Flume clients when we started TD migration. The gRPC Chubby fallback counter was sensitive enough to detect and alert this which we tentatively concluded TD overload/contention leading to timeouts, so the failure was gray. I think a gauge could have shown something eventually but certainly would have been much less sensitive in this case as they were brief and gray.

Reply:
Can't we detect these cases by looking at server-side statistics on the fallback server?  If any client is talking to the fallback server, we know it wasn't able to use the primary.

I'm not strongly opposed to adding a counter if we really need it.  I just want to be convinced that we can't actually get what we need from the existing data.

Reply:
Likely for the primary failure but not for the fallback failure. From what we saw so far , I expect backup failure can happen if the control plane server fails to keep up with the growth. When it happens I think only the client side counter type can really tell us the actual magnitude of the impact.

Reply:
A fallback counter wouldn't help to catch cases where the fallback server is unavailable, because there's nothing to fall back to from there.  It sounds like what you actually want here is a counter of the number of times the channel for a given server goes from healthy to unhealthy.  Is that what you want?

Reply:
I've added a proposed counter for this.  PTAL.

Reply:
Does it count as a server failure when the currently active subchannel fails but it could switch to another successfully? If not, this LGTM.

Reply:
No, it won't.  As per A57, we consider it failing only when either (a) the channel reports TRANSIENT_FAILURE or (b) the ADS stream fails without the client having seen any response messages.